src/nfa/intersection.cc

```diff
@@ -16,177 +16,254 @@
+    std::deque<State> pairs_to_process{}; // Set of state pairs of original states to process.
+
+    //The largest matrix (product_matrix) of pairs of states we are brave enough to allocate.
+    // Let's way we are fine with allocating large_product * (about 8 Bytes) space.
+    // So ten million cells is close to 100 MB.
+    // If the number is larger, then we do not allocate a matrix, but use a vector of unordered maps (product_vec_map),
+    // together with a heuristic that speeds up containment test, because unordered map lookup is not cheap:
+    // We remember for every lhs/rhs state a range of rhs/lhs states that have already appeared with it (vectors min_lhs/rhs and max_lhs/rhs),
+    // And every containment test first asks whether lhs and rhs are in each others ranges.
+    // This is several times faster compared to pure product_vec_map, which is turn is notably faster that one unordered map from pairs to states.
+    // (but Juraj says that rewriting hash function may help unordered map significantly, so maybe that would be enough ... ?)
+    // TODO: where to put this magical constant? It should not be here.
+    const bool large_product = lhs.num_of_states() * rhs.num_of_states() > 100000000;
+
+    ProductMatrix product_matrix;
+    ProductVecMap product_vec_map;
+    std::vector<State> min_rhs;
+    std::vector<State> max_rhs;
+    std::vector<State> min_lhs;
+    std::vector<State> max_lhs;
+
+    auto update_ranges = [&min_rhs,&max_rhs,&min_lhs,&max_lhs](State lhs_state, State rhs_state)
+    {
+        if (min_rhs[lhs_state] > rhs_state)
+            min_rhs[lhs_state] = rhs_state;
+        if (max_rhs[lhs_state] < rhs_state)
+            max_rhs[lhs_state] = rhs_state;
+        if (min_lhs[rhs_state] > lhs_state)
+            min_lhs[rhs_state] = lhs_state;
+        if (max_lhs[rhs_state] < lhs_state)
+            max_lhs[rhs_state] = lhs_state;
+    };
+
+    auto are_in_range = [&min_rhs,&max_rhs,&min_lhs,&max_lhs](State lhs_state, State rhs_state) {
+        return (rhs_state <= max_rhs[lhs_state] && rhs_state >= min_rhs[lhs_state] && lhs_state <= max_lhs[rhs_state] && lhs_state >= min_lhs[rhs_state]) ;
+    };
+
+    //Initialize the storage, according to the number of possible state pairs.
+    if (!large_product)
+        product_matrix = ProductMatrix(lhs.num_of_states(), std::vector<State>(rhs.num_of_states(), Limits::max_state));
+    else {
+        product_vec_map = ProductVecMap(lhs.num_of_states());
+        min_rhs = std::vector<State>(lhs.num_of_states(), Limits::max_state);
+        max_rhs = std::vector<State>(lhs.num_of_states(), Limits::max_state);
+        min_lhs = std::vector<State>(rhs.num_of_states(), 0);
+        max_lhs = std::vector<State>(rhs.num_of_states(), 0);
+    }
+
+    auto product_contains = [&product_matrix,&product_vec_map,&large_product,&are_in_range](State lhs_state, State rhs_state) {
```

Comment:
Simplify lambda capture
```suggestion
    auto product_contains = [&](State lhs_state, State rhs_state) {
```

Reply:
done